core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -384,40 +355,62 @@ public CompletableFuture<Map<TopicIdPartition, ShareAcknowledgeResponseData.Part
+        shareFetchData.forEach((tp, sharePartitionData) -> {
+            if (sharePartitionData.maxBytes > 0) shareFetchDataWithMaxBytes.put(tp, sharePartitionData);
+        });
         if (reqMetadata.isFull()) {
-            String removedFetchSessionStr = "";
             // TODO: We will handle the case of INVALID_MEMBER_ID once we have a clear definition for it
          //  if (!Objects.equals(reqMetadata.memberId(), ShareFetchMetadata.INVALID_MEMBER_ID)) {
             ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
-            if (cache.remove(key) != null)
-                removedFetchSessionStr = "Removed share session with key " + key;
-
+            String removedFetchSessionStr = "";
             if (reqMetadata.epoch() == ShareFetchMetadata.FINAL_EPOCH) {
                 // If the epoch is FINAL_EPOCH, don't try to create a new session.
-                context = new FinalContext(shareFetchData);
+                if (!shareFetchDataWithMaxBytes.isEmpty()) {
+                    throw Errors.INVALID_REQUEST.exception();
+                }
+                context = new FinalContext();
+                if (cache.remove(key) != null) {
+                    removedFetchSessionStr = "Removed share session with key " + key;
+                    log.debug(removedFetchSessionStr);
+                }
             } else {
-                context = new ShareSessionContext(time, cache, reqMetadata, shareFetchData);
+                if (cache.remove(key) != null) {
+                    removedFetchSessionStr = "Removed share session with key " + key;
+                    log.debug(removedFetchSessionStr);
+                }
+                ImplicitLinkedHashCollection<CachedSharePartition> cachedSharePartitions = new
+                        ImplicitLinkedHashCollection<>(shareFetchDataWithMaxBytes.size());
+                shareFetchDataWithMaxBytes.forEach((topicIdPartition, reqData) -> {
+                    cachedSharePartitions.mustAdd(new CachedSharePartition(topicIdPartition, reqData, false));
+                });
+                ShareSessionKey responseShareSessionKey = cache.maybeCreateSession(groupId, reqMetadata.memberId(),
+                        time.milliseconds(), shareFetchDataWithMaxBytes.size(), cachedSharePartitions);
+                log.debug("Share session context with key {} isSubsequent {} returning {}", responseShareSessionKey,
+                        false, partitionsToLogString(shareFetchDataWithMaxBytes.keySet()));
+
+                context = new ShareSessionContext(time, cache, reqMetadata, shareFetchDataWithMaxBytes);
                 log.debug("Created a new ShareSessionContext with {} {}. A new share session will be started.",
-                        partitionsToLogString(shareFetchData.keySet()), removedFetchSessionStr);
+                        partitionsToLogString(shareFetchDataWithMaxBytes.keySet()));
```

Comment:
The number of placeholders do not match number of arguments mentioned in the log statement